rewrite-java/src/main/java/org/openrewrite/java/tree/J.java

```diff
@@ -363,6 +367,58 @@ public CoordinateBuilder.Expression getCoordinates() {
         }
     }
 
+    @FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+    @EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
+    @Data
+    final class ArrayTypeTree implements J, TypeTree, Expression {
+        @With
+        @EqualsAndHashCode.Include
+        UUID id;
+
+        @With
+        Space prefix;
+
+        @With
+        Markers markers;
+
+        @With
+        TypeTree elementType;
+
+        @Nullable
+        @With
+        List<J.Annotation> annotations;
```

Comment:
With an `AnnotatedType` like the compiler, we'd need to check the parent cursor from `visitArrayType` before printing the dimensions. I'm not in favor of `J.AnnotatedType` => `J.ArrayType`, since it is pretty odd, and I'm unaware of any other elements that work that way. The compiler has a `pos` field on annotations, so, I think it's different since we can't maintain a pos on annotations in the LST. Do you think there is a significant benefit to wrapping the `J.ArrayType` with a `J.AnnotatedType`?

### Comparison of `int @A1 [] @A2 []`.

### `J.AnnotatedType` to `J.ArrayType`
J.AnnotatedType:
- annotations: `@A2` 
- typeExpression:
    - J.ArrayType:
        - elemType:
            - J.AnnotatedType:
                - annotations: `@A1`
                - typeExpression:
                    - J.ArrayType:
                        - elemType:
                            - J.Identifier: `int`
                        - dimension: `[]`
        - dimension: `[]`

vs

### `J.ArrayType` with annotations field
J.ArrayType:
- elemType:
    - J.ArrayType:
        - elemType:
            - J.Identifier: `int`
        - annotations: `@A1`
        - dimension: `[]`
- annotations: `@A2`
- dimension: `[]`

Reply:
As far as the printing is concerned I would instead do it the other way around: When visiting an annotated type check if it decorates an array type and then inline some of its code. But that isn't very pretty either.

The only concern I have is that recipes need to be aware of another location where annotations can be defined. But the AnnotationService might help here.

I also think we should probably leave it as it is for now. We can still re-evaluate if we discover problems.